[Java] Kyma On-Prem Connectivity Workaround (#954)
SAP/cloud-sdk: docs-java/environments/sap-btp-kyma-environment.mdx

We are currently re-evaluating and updating the information provided.
Please follow the [official documentation](https://kyma-project.io/docs/kyma/latest/01-overview/).

<details>
  <summary>Workaround for On-Premise Connectivity</summary>

> Some of the most recent changes of the Kyma runtime environment render the SAP Cloud SDK's On-Premise connectivity unfunctional.

Author: ```suggestion
Recent changes of the Kyma runtime environment prevent the SAP Cloud SDK's On-Premise connectivity from working as intended.
```

While we are working on a permanent solution, please use the following workaround:

<ol>
  <li>Add following class to your project:

```java
public class OnPremiseHeaderProvider implements DestinationHeaderProvider
{
    @Override
    @Nonnull
    public List<Header> getHeaders( @Nonnull final DestinationRequestContext requestContext )
    {
        final Option<ProxyType> proxyType = requestContext.getDestination().get("ProxyType", ProxyType.class);
        if( proxyType.getOrElse(ProxyType.INTERNET) != ProxyType.ON_PREMISE ) {
            return Collections.emptyList();
        }

        final Option<AuthenticationType> authentication =
            requestContext.getDestination().get("authentication", AuthenticationType.class);

>         if( authentication
>             .getOrElse(AuthenticationType.NO_AUTHENTICATION) != AuthenticationType.PRINCIPAL_PROPAGATION ) {

Reviewer: Maybe slightly less code...?
```suggestion
        if( !authentication.contains(AuthenticationType.PRINCIPAL_PROPAGATION) ) {
```

>             return Collections.emptyList();

Reviewer 2: Why don’t we cover principal propagation? Is it specific to the untrusted mode?

        }

        final Try<String> maybeBearerToken =
            AuthTokenAccessor
                .tryGetCurrentToken()
                .map(token -> String.format("Bearer %s", token.getJwt().getToken()));

        if( maybeBearerToken.isEmpty() ) {

>             return Collections.emptyList();

Reviewer 2: At this point a warning might make sense, not sure if there is a scenario where the onprem call afterwards could still succeed
